### ISSUE TITLE
Callback Functions Support

### ISSUE THREAD

**Issue by bo-ke** [NVIDIA-NeMo/Megatron-Bridge#783]
Will you support callback functions such as on_load_data_end and on_optimizer_begin, similar to the callbacks in Hugging Face?"

**Comment by ananthsub** [comment]
Hi @bo-ke , could you describe what sort of customizations you'd need to run in these hooks? 

the initial design for the training loop was without callbacks based on previous feedback that programming based on callbacks can add significant overhead

**Comment by bo-ke** [comment]
> Hi [@bo-ke](https://github.com/bo-ke) , could you describe what sort of customizations you'd need to run in these hooks?
> 
> the initial design for the training loop was without callbacks based on previous feedback that programming based on callbacks can add significant overhead

What I needs are as follows:
Add additional logging when on_step_end is triggered.
Conduct an analysis of the parameter's  gradients right before optimizer_step runs.
Calculate and record the source distribution of the training data once on_load_data_end is executed.
With a callback mechanism in place, I can inject my custom code conveniently without having to modify the original megatron_bridge code.

To add to this, the current implementation of the train_logging code is too rigid. For instance, if I want to inject additional keys into TensorBoard for logging, I have no choice but to modify (hack) the existing code directly.

**Comment by yaoyu-33** [comment]
Just FYI, we are working on implementing the hooks now. The support will come in next month or so.

**Comment by BlueCrescent** [comment]
In this context, it would be very useful to have a callback option after writing out a checkpoint is finished (in order to trigger tasks like downstream evaluation or checkpoint upload/transfer).

**Comment by yaoyu-33** [comment]
callbacks implemented https://docs.nvidia.com/nemo/megatron-bridge/latest/training/callbacks.html
